### PR TITLE
Basic Type Refinement Visualization for IDE Plugins

### PR DESCRIPTION
**Pull request by zcai1**
This PR writes type refinement of local variables to javac's diagnostics listener. To enable this feature, simply pass `-AprintTypeRefinements` as a command line option to checker framework.

In stdout, the type refinement information will be printed in the following format:
```
Sample1.java:30: Note: [type.refinements] variable="s", types=[@UnknownKeyFor]
        var s = new Sample1();
            ^
Sample1.java:30: Note: [type.refinements] variable="s", types=[@Initialized, @NonNull]
        var s = new Sample1();
            ^
Sample1.java:31: Note: [type.refinements] variable="s", types=[@UnknownKeyFor]
        System.out.println(s.foo().toString());
                           ^
Sample1.java:31: Note: [type.refinements] variable="s", types=[@Initialized, @NonNull]
        System.out.println(s.foo().toString());
                           ^
```

In [CF Language Server](https://github.com/eisopux/checker-framework-languageserver), there's a [JavacWrapper](https://github.com/eisopux/checker-framework-languageserver/blob/master/src/main/java/org/checkerframework/languageserver/JavacWrapper.java) which collects all diagnostics from checker framework and parse them into [CFDiagnostic](https://github.com/eisopux/checker-framework-languageserver/blob/master/src/main/java/org/checkerframework/languageserver/CFDiagnostic.java). 

To differentiate type refinement from other diagnostics information, there are two properties we can check:
1. `CFDiagnostic.kind == "NOTE"`
2. `CFDiagnostic.message == "variable="...", types=[...]"`

Currently, all diagnostics information will be published to client, so we can easily obtain some visulization examples:
image
image

Notes:
1. `CFDiagnostic.lineNumber` and `CFDiagnostic.columnNumber` are the line and column number of the start position of a local variable's identifier.
2. Since identifier can't be splitted into multiple lines, the end position will have the same line number. However, our current way of calculating the column number of end position in [CFTextDocumentService](https://github.com/eisopux/checker-framework-languageserver/blob/66c90b22b19d174c1be5c0a5301bcf741b3910fb/src/main/java/org/checkerframework/languageserver/CFTextDocumentService.java#L71) will not be accurate for type refinement. This is because javac will somehow include the rhs of the assignment tree, even if only lhs is provided in the diagnostic message. To calculate the correct column number of end position, we can compute `endCol = startCol + length(variableName)`.

### CODE REVIEW

**Comment by zcai1** [comment]
We can consider stdout instead of txt files.

**Comment by zcai1** [comment]
Cherry-picked https://github.com/opprop/checker-framework/pull/179/commits/17814cd6470204925f73a8572fa9c36eefdf40cc to allow publishing build artifacts. The commit should be reverted before merging.

File: framework/src/main/java/org/checkerframework/framework/util/TypeRefinementVisualizer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.checkerframework.framework.flow.CFAbstractStore;
+import org.checkerframework.framework.flow.CFAbstractValue;
+import org.checkerframework.framework.source.DiagMessage;
+import org.checkerframework.framework.source.SourceChecker;
+
+import javax.tools.Diagnostic;
+
+public class TypeRefinementVisualizer {
+
+    public void visualize(
```
Comment (zcai1) [inline framework/src/main/java/org/checkerframework/framework/util/TypeRefinementVisualizer.java:19]: Should we make it a static method?

**Comment by zcai1** [comment]
added checker name to logs, e.g.,
```
Sample1.java:30: Note: [type.refinements] variable="s", checker="KeyFor", types=[@UnknownKeyFor]
        var s = new Sample1();
            ^
Sample1.java:30: Note: [type.refinements] variable="s", checker="Nullness", types=[@Initialized, @NonNull]
        var s = new Sample1();
            ^
```

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -1326,8 +1331,12 @@ protected void analyze(
      *     com.sun.source.tree.ClassTree, boolean, boolean, boolean,
      *     org.checkerframework.framework.flow.CFAbstractStore)
      */
-    protected void postAnalyze(ControlFlowGraph cfg) {
+    protected void postAnalyze(ControlFlowGraph cfg, AnalysisResult<Value, Store> analysisResult) {
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:1334]: Overall, can you try putting this logic into https://github.com/opprop/checker-framework/blob/08e2ea45ea847a832d05ebb06b1128675ab332f8/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java#L1139 ?
There, define an AST visitor that outputs information for the interesting nodes. This shouldn't just be type refinements, but also method invocations, binary operations, etc. Basically anything that can have an annotated type.

File: framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java [hunk 3]
```diff
@@ -348,6 +348,11 @@
     // -Acfgviz=MyClass,verbose
     "verbosecfg",
 
+    /// LSP Support
+
+    // Whether to print refined types for local variables
+    "printTypeRefinements",
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java:354]: How about `outputAnnotatedTypes` to reflect that this isn't just for type refinements?
Or `outputImplicitAnnotations`, to indicate it's not just type annotations? ("Defaulted" would be too limiting, but "Implicit" is maybe also not the best term. Maybe "outputEffectiveAnnotations"?)

File: framework/src/main/java/org/checkerframework/framework/source/messages.properties [hunk 4]
```diff
@@ -1,2 +1,3 @@
 annotation.not.completed=Element %s contains annotation %s which can't be found. Make sure your classpath is set correctly.
 unneeded.suppression=warning suppression %s is not used by %s
+type.refinements=variable="%s"; checker="%s"; types=%s
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/source/messages.properties:3]: For the key, just use singular and adapt to the suggestion on the option name.
For the value, in the medium term, I wouldn't limit this to variables. Maybe nothing is needed, as it just gives the type for a particular range.

File: framework/src/main/java/org/checkerframework/framework/util/TypeRefinementVisualizer.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                if (value == null) continue;
+
+                checker.report(
+                        localVariableNode.getTree(),
+                        new DiagMessage(
+                                Diagnostic.Kind.OTHER,
+                                "type.refinements",
+                                localVariableNode.getName(),
+                                checkerName,
+                                value.getAnnotations()));
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/TypeRefinementVisualizer.java:54]: When you moved this logic, the output should be the AnnotatedTypeMirror for the tree, not just the primary annotations.

**Review by wmdietl — COMMENTED** [review]
A few comments, but it's easier to chat in our meeting.

**Pull request closed without merging**